### DISCUSSION TITLE
Getting base type out of datamodel

### DISCUSSION THREAD

**Discussion by jtniehof** [spacepy/spacepy#719] [Design and roadmapping]
tl;dr: what should be the API for getting `ndarray` out of `dmarray` (and `dict` out of `SpaceData`)?

Unfortunately the subclassing of `numpy.ndarray` to `dmarray` results in some weird side-effects so explicit casts back down to `ndarray` are still necessary.

There was agreement back at the [first PyHC meeting](https://docs.google.com/document/d/1ejP0kaibf6DRxjYJNmPrF1t3Nl21r0pC1FuDhu0hPnM/edit#heading=h.vewcpo8g2yd4) that a `.data` attribute should return the base object. (There's also some interesting stuff on disovering when you're being called in an array context that I think we work around by subclassing, but would be an alternative....) Unfortunately we can't do this for a `dmarray` because [.data is already defined for ndarray](https://numpy.org/devdocs/reference/generated/numpy.ndarray.data.html) and thus it's not available for use in dmarray unless we break ndarray compatibility in that way.

`ndarray` has methods `.tolist()`, `.tobytes()`, etc. so it might make sense for `dmarray` to have `.toarray()` or `.tondarray()`...which one of those makes more sense is another question. (`numpy.asarray` seems to suggest using `array` even though `ndarray` is technically correct).

By analogy then `SpaceData` could have `.todict()`, which has the downside of not being the _same_ as `dmarray`.

As a list of options:

1. `toarray()` / `tondarray()` (with `.todict()`)
2. Overload `.data` so that you'd need `.data.data` to actually get the underlying buffer
3. `tobase()` or something simlar on both `dmarray` and `SpaceData`
4. as always, something else....
5. nothing, keep with `numpy.asarray(foo)` where `foo` is a `dmarray`.

**Comment by jtniehof** [comment]
Context on dmarray difficulties: #13 attempted reimplementation to fix them; #12, #258, #528.

**Comment by spacecataz** [comment]
My preferences, in order:

#1, but using `toarray()` for brevity/simplicity.
#2, because the `.data` syntax for base ndarrays is not really useful ever.
#5. This one is my third option because the method is convenient and tab-completeable in IPython.

In any case, it would be great to attack the underlying issue again, but I get that this is tough.

**Reply by jtniehof** [reply]
I never use `.data` myself but I wonder how many things inside numpy use it and would blow up if we implement this...

**Comment by jtniehof** [comment]
astropy Quantities uses `.value`. Pandas uses `.values`. :facepalm:
